Read-only barometer in Position_estimation

jlecoeur opened the pull request:
Position_estimation has a non const reference to `Barometer`, and it calls `barometer.calibratre_bias(...)` which modifies the barometer object.

This is an issue when the barometer is used in another filter (say a Kalman) in parallel to Position_estimation. When Position_estimation decides to updates the barometer bias, the Kalman sees a jump of several meters on the barometer value.

This PR updates position_estimation in order to use its own bias on the barometer.

jlecoeur commented:
Now the question is whether we should keep the methods `Barometer:calibrate_bias()` and `Barometer::has_been_calibrated()`.
As you worked on this @basil-huber and @mdouglas90,  what do you think ?

mdouglas90 commented:
We probably shouldn't have two separate calibrations accessible from the same place because someone wouldn't easily know which one to use, So if we are moving it to the Position_estimation class, it should be removed from the Barometer class. But it would still be better to have it somehow only in the Barometer class.

Should the Kalman filter be updating itself from the barometer when it has not been calibrated?

jlecoeur commented:
The kalman also estimates the barometer bias, it does not really matter if the barometer is not calibrated.

It would be fine if it was calibrated once (though every filter would have to check if the barometer is calibrated, and do it if it is not). 

The big problem is that Position_estimation recalibrates the baro every time we arm, so just before flight!

jlecoeur commented:
Tested today outside, Position_estimation was behaving well with this change.

jlecoeur commented:
@GregoireH @mdouglas90 what do you think ?
I removed the calibration methods in Barometer interface. But I added a member 
```cpp
static float Barometer::pressure_at_sea_level_;
```
and associated static methods  
```cpp
static float Barometer::altitude_from_pressure(float pressure, float pressure_at_sea_level = pressure_at_sea_level());
static float Barometer::compute_pressure_at_sea_level(float pressure, float altitude);
```

The behaviour is unchanged compared to after the first commit (when I last tested), because `Barometer::calibrate_bias(float current_altitude_gf)` was never called.
Just now we have the additional capability to calibrate the pressure at sea level. The pressure at sea level is used to convert pressure to altitude.

GregoireH commented on drivers/barometer.cpp:
```diff
@@ -48,25 +48,34 @@ extern "C"
-    has_been_calibrated_(false)
-{
+// Definition of static member
+float Barometer::pressure_at_sea_level_;
```
why is this here ? and not in the class definition instead ?

jlecoeur replied:
Same as `INS::origin_`. This is static, so the variable needs to be defined somewhere

GregoireH commented on drivers/barometer.cpp:
```diff
@@ -48,25 +48,34 @@ extern "C"
 {
-    altitude_bias_gf_ = altitude_gf() - current_altitude_gf;
-    has_been_calibrated_ = true;
+    return 44330.0f * (1.0f - pow(pressure / pressure_at_sea_level, 0.190295f));
```
@jlecoeur 
Could you add a link to where you get the formula from please ?

jlecoeur replied:
https://en.wikipedia.org/wiki/Atmospheric_pressure

GregoireH commented on drivers/barometer.cpp:
```diff
@@ -48,25 +48,34 @@ extern "C"
+float Barometer::compute_pressure_at_sea_level(float pressure, float altitude)
 {
-    return 44330.0f * (1.0f - pow(pressure / 101325.0f, 0.190295f)) - altitude_bias;
+    return pressure / pow(1.0f - altitude / 44330.0f, 5.255);
```
missing an f for 2nd arg isn't it ?

jlecoeur replied:
done in 5d50e3a43a747ad6af1d240e61191d93d2dcc2cb

GregoireH commented on drivers/barometer.hpp:
```diff
@@ -116,36 +116,48 @@ class Barometer
-    float altitude_bias_gf_;    ///< Offset of the barometer sensor for matching GPS altitude value
-    bool has_been_calibrated_;  ///< Flag stating if the barometer has been calibrated
+private:
+    static float pressure_at_sea_level_;    ///< Pressure at sea level. This is a static variable shared by all barometers
```
could you add the unit please ?

jlecoeur replied:
done in https://github.com/lis-epfl/MAVRIC_Library/pull/311/commits/f20f542ceab4c749dbeae84ed50c237d36f02cb7

GregoireH commented on drivers/barometer_bmp085.cpp:
```diff
@@ -250,7 +249,7 @@ bool Barometer_BMP085::update(void)
 
             // remove bias
-            altitude_gf_ = altitude_filtered - altitude_bias_gf_;
+            altitude_gf_ = altitude_filtered;
```
do you still need to have those 2 variables then ?

jlecoeur replied:
done in https://github.com/lis-epfl/MAVRIC_Library/pull/311/commits/abad5bd90f89aa8de0fdcd1a8bfed8e493b2ec32

GregoireH commented on libs/mavlink/include:
```diff
@@ -1 +1 @@
-Subproject commit 0a6ec5d8f7788b38b1a76a49136dad72fa750fe9
+Subproject commit 6f36e00c3862a068a460dd169a0118eff5b68412
```
I suspect that it comes from a commit -am isn't it ?

jlecoeur replied:
yes must be :-)

jlecoeur replied:
https://github.com/lis-epfl/MAVRIC_Library/pull/311/commits/5ea28080aed6eb7fd0cdba366a53c16b5f22b85b

GregoireH commented on sample_projects/LEQuad/main_avr32.cpp:
```diff
@@ -72,7 +71,7 @@ int main(void)
     // Create board
     // -------------------------------------------------------------------------
     megafly_rev4_conf_t board_config    = megafly_rev4_default_config();
-    board_config.imu_config             = imu_config();                         // Load custom imu config (cf conf_imu.h)
```
if this is not releavant anymore we could remove it instead of commenting

jlecoeur replied:
uncommented in https://github.com/lis-epfl/MAVRIC_Library/pull/311/commits/674fc5fd12fdc1576ff7ea165e6324663a3e1443

GregoireH commented on sample_projects/LEQuad/proj_avr32/config/conf_imu.hpp:
```diff
@@ -48,59 +48,59 @@ extern "C"
 }
 
 ///< Define which configuration of the imu to use, depending on the autopilot ID
-#if MAVLINK_SYS_ID == 1
```
idem

jlecoeur replied:
This is relevant as an example of how to do if you have several imu configs.

GregoireH commented on sample_projects/LEQuad/proj_avr32/config/conf_platform.h:
```diff
@@ -46,7 +46,7 @@
 #endif
 
-#define MAVLINK_SYS_ID 0
+#define MAVLINK_SYS_ID 2
```
add a note so that people don't decide to set it back to 0 nor 1 :) please

jlecoeur replied:
done in https://github.com/lis-epfl/MAVRIC_Library/pull/311/commits/154be9e70e2c85dfc062b9c418bdc1fd48ee1bdb

GregoireH commented on sensing/position_estimation.hpp:
```diff
@@ -201,10 +201,13 @@ class Position_estimation: public INS
 
     float gravity;                          ///< Value of the gravity
 
+    float barometer_bias;
```
missing doc

jlecoeur replied:
done in https://github.com/lis-epfl/MAVRIC_Library/pull/311/commits/2f813793cac5e2ebf2a7fd5f365261eca57dc003

GregoireH commented on sensing/position_estimation.hpp:
```diff
@@ -201,10 +201,13 @@ class Position_estimation: public INS
     State& state;                           ///< Reference to the state structure
     const Gps& gps;                         ///< Reference to the GPS structure
-    Barometer& barometer;                   ///< Reference to the barometer structure
+    const Barometer& barometer;                   ///< Reference to the barometer structure
```
alignment

jlecoeur replied:
done in https://github.com/lis-epfl/MAVRIC_Library/pull/311/commits/2f813793cac5e2ebf2a7fd5f365261eca57dc003
